### PR TITLE
libiwl: phase 1 modernization (collapse dup, overflow check, RAII)

### PR DESCRIPTION
**Pull request by susilehtola**
Phase 1 of the libiwl modernization tracked in #3425. An internal refactor of `psi4/src/psi4/libiwl/` with **no caller-facing API change and no on-disk format change**, plus a standalone round-trip test for the bucket I/O. The ~30 in-tree consumers that poke `struct iwlbuf` fields directly all keep compiling unchanged.

## Description
The libiwl bucket reader/writer had two byte-identical bodies for every operation (a C `iwl_buf_*` free function and a C++ `IWL::` method), C-style macros, mixed `malloc`/`new[]` ownership, and silent 32-bit→16-bit narrowing at the `Label` cast that could corrupt a bucket file without warning. This PR consolidates the duplication, fixes the latent correctness/ownership issues, and adds a round-trip test — all without touching the public field layout or the on-disk format.

## User API & Changelog headlines
- [x] `Label` overflow (orbital index exceeding the 16-bit bucket field) now throws a `PSIEXCEPTION` naming the offending indices instead of silently narrowing and corrupting the IWL file.

## Dev notes & details
- `config.h`: replace `#define IWL_INTS_PER_BUF` / `IWL_KEY_*` with `inline constexpr`; document the bucket layout.
- `iwl.hpp`: remove dead declarations `IWL::read` / `IWL::to_end` (never defined); delete copy ctor / assign; document the deferred-init pattern used by `libtrans`.
- Collapse every duplicated C-function / C++-method body into a single shared static helper per operation (one implementation site instead of two byte-identical ones). New internal `iwl_impl.h` holds the shared `check_label_fits()` / `make_printer()` helpers.
- **Overflow check at every `(int) → Label` cast** in the three write paths (throws instead of silently narrowing).
- `buf_init`: throw on `tocscan` failure instead of `Printf`'ing and leaving a half-initialized buffer the caller then `fetch()`es from.
- Unify allocators on `new[]`/`delete[]` (removes the cross-API `malloc`/`free` vs `new[]`/`delete[]` ownership UB).
- Hoist per-integral `PsiOutStream` construction in the write paths to one per-call instance, only when `printflag` is set (previously one file open per nonzero integral).
- Default-construct all `IWL` members including `keep_`; drop `MAX0`/`MIN0`/`INDEX` macros in `buf_wrt_mat.cc` for `std::max`/`std::min` and an inline `packed_index()`.

**Test.** `test/iwl_roundtrip_test.cc` is a standalone round-trip exerciser covering the bucket-boundary cases the 1995 README warned about — empty, 1, `B-1`, `B`, `B+1`, `3.5*B` integrals — plus the `Label` overflow check. It is built and registered as the `libiwl-roundtrip` ctest (`tests/libiwl/`, in the superbuild where `enable_testing()` runs), labeled `psi;quicktests;smoketests;libiwl`. It links the libiwl/libpsio/libpsi4util/liboptions/libciomr archives plus a tiny stubs TU (`test/stubs.cc`) that supplies the runtime globals normally provided by `core.cc`/libdpd (`global_dpd_`, `psi_file_prefix`, a no-op `C_DSYEV`), so it needs neither Python nor libdpd. It is skipped on Windows, where the curated link line does not resolve under clang-cl/lld-link (see `psi4/src/psi4/libiwl/CMakeLists.txt`).

## What this PR does *not* do
- No change to `struct iwlbuf`'s field layout — direct field pokes keep compiling.
- No on-disk format change; the v1 bucket layout (`int lastbuf | int inbuf | Label[4*2980] | Value[2980]`) is preserved bit-for-bit.
- The hard 16-bit orbital-index cap is unchanged; only the failure mode changes (loud throw vs silent corruption). Bumping to 32-bit is phase 3.
- Phase 2 (iterator-based `IWLReader`/`IWLWriter` + caller migration, #3439–#3441) and phase 3 (versioned wire format, async default, optional compression) are follow-up PRs. See #3425.

## Questions
- [ ] None outstanding.

## AI
- [x] AI usage: refactor, the round-trip test and its hosting stubs, and bug-seeking (the `Label` overflow and allocator-mismatch issues).

## Checklist
- [x] Tests added for any new features
- [ ] Docstring or narrative docs/sphinxman/source updated if needed (internal refactor; none needed)
- Test running is well covered by CI. For running locally, see [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- Ready for merge -- use Draft/Open GitHub status to signal your view on merge readiness

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by susilehtola** [comment]
~~Heads up: #3428 contains this phase-1 work **plus** the rest of the libiwl modernization (the `IWLReader`/`IWLWriter` API, the full consumer fan-out, removal of the dead C API, and signature cleanups). If you'd prefer to review the whole effort in one place, #3428 is a superset of this PR and this one can be closed; if you'd rather land it incrementally, this remains the first slice. Either way works for me.~~

**Superseded.** #3428 turned out to be too large to review and has been closed. The incremental route won: this PR is the base of the stack, followed by #3439 → #3440 → #3441.

File: psi4/src/psi4/libiwl/iwl.hpp [hunk 1]
```diff
@@ -30,33 +30,38 @@
...
+    int ints_per_buf_;
+    int bufszc_;
+    double cutoff_;
+    int lastbuf_;
+    int inbuf_;
+    int idx_;
+    Label *labels_;
+    Value *values_;
     PSIO *psio_;
-    /*! Flag indicating whether to keep the IWL file or not */
```
Comment (TiborGY) [inline psi4/src/psi4/libiwl/iwl.hpp:51]: Why did you remove these comments?

**Comment by TiborGY** [comment]
This PR has accumulated a long and messy history. If possible I would prefer for the new tests to go in first as a new PR, and then what remains of this can be respun as a new PR on top of that.

**Comment by susilehtola** [comment]
Superseded by the split you asked for, @TiborGY:

- #3465 — the round-trip test, landed first against current libiwl (no library changes), so it guards the refactor.
- #3466 — the phase-1 modernization respun on top of #3465, with a clean history.

The test's `Label`-overflow case and the `psi::IWL_INTS_PER_BUF` constexpr necessarily live in #3466 (they test/require the new behavior); the behavior-preserving round-trip cases are in #3465. I also restored the `IWL` field comments in #3466. Closing this in favor of those two.
